client-sdk/go/types/crypto.go

```diff
@@ -71,3 +72,69 @@ func (pk *PublicKey) UnmarshalJSON(data []byte) error {
 	}
 	return pk.unmarshal(&spk)
 }
+
+// MultisigSigner is one of the signers in a multisig configuration.
+type MultisigSigner struct {
+	PublicKey PublicKey `json:"public_key"`
+	Weight    uint64    `json:"weight"`
+}
+
+// MultisigConfig is a multisig configuration.
+// A set of signers with total "weight" greater than or equal to a "threshold" can authenticate
+// for the configuration.
+type MultisigConfig struct {
+	Signers   []MultisigSigner `json:"signers"`
+	Threshold uint64           `json:"threshold"`
+}
+
+// Verify performs some sanity checks.
+func (mc *MultisigConfig) Verify() error {
+	if mc.Threshold == 0 {
+		return fmt.Errorf("zero threshold")
+	}
+	var total uint64
+	encounteredKeys := make(map[PublicKey]bool)
+	for i, signer := range mc.Signers {
+		if encounteredKeys[signer.PublicKey] {
+			return fmt.Errorf("signer %d duplicated", i)
+		}
+		encounteredKeys[signer.PublicKey] = true
+		if signer.Weight == 0 {
+			return fmt.Errorf("signer %d zero weight", i)
+		}
+		newTotal, carry := bits.Add64(total, signer.Weight, 0)
+		if carry != 0 {
+			return fmt.Errorf("weight overflow")
+		}
+		total = newTotal
+	}
+	if total < mc.Threshold {
+		return fmt.Errorf("impossible threshold")
+	}
+	return nil
+}
+
+// Batch checks that enough signers have signed and returns vectors of public keys and signatures
+// for batch verification of those signatures. This internally calls `Verify`.
+func (mc *MultisigConfig) Batch(signatureSet [][]byte) ([]PublicKey, [][]byte, error) {
```

Comment:
Why not do the actual verification?  I'm of mixed opinion if a batch should fail if it as any invalid signatures regardless of threshold or not.  The way this is setup right now is an all-or-nothing design as far as I can tell.

(I assume allowing a MultisigConfig with mixed public-key types is deliberate, though this will complicate batch verification logic.)

Reply:
we'll have to have an additional step to separate the public key types. but are we moving away from batch signature verification in general?

this isn't meant to be all-or-nothing. it'll leave out the `nil`s from the signature set (Option<Signature> in the rust side). we'd advise that the transaction sender nil out invalid signatures on their own and avoid submitting a transaction with some valid and some invalid signatures

Reply:
I'll open an issue to discuss this

Reply:
#143